File: fronted/src/api/tag.js
import request from '../utils/request'

// 获取标签列表
export function getTagList(data) {
  return request({
    url: '/tag/getList',
    method: 'post',
    data
  })
}

// 获取单个标签
export function getTag(data) {
  return request({
    url: '/tag/getOne',
    method: 'post',
    data
  })
}

// 创建标签
export function createTag(data) {
  return request({
    url: '/tag/create',
    method: 'post',
    data
  })
}

// 更新标签
export function updateTag(data) {
  return request({
    url: '/tag/update',
    method: 'post',
    data
  })
}

// 删除标签
export function deleteTag(data) {
  return request({
    url: '/tag/delete',
    method: 'post',
    data
  })
}

// 通过 slug 获取标签
export function getTagBySlug(data) {
  return request({
    url: '/tag/getBySlug',
    method: 'post',
    data
  })
}

File: backend/internal/service/tag.go
package service

import (
	"context"

	"blog/internal/dao"
	"blog/internal/model/entity"
)

type sTag struct{}

var Tag = sTag{}

func (s *sTag) Create(ctx context.Context, in *entity.Tag) (id int, err error) {
	result, err := dao.Tag.Ctx(ctx).Data(in).Insert()
	if err != nil {
		return 0, err
	}
	id64, _ := result.LastInsertId()
	return int(id64), nil
}

func (s *sTag) Update(ctx context.Context, in *entity.Tag) (err error) {
	_, err = dao.Tag.Ctx(ctx).Data(in).Where(dao.Tag.Columns().Id, in.Id).Update()
	return
}

func (s *sTag) GetList(ctx context.Context, name, slug string, page, size int) (list []*entity.Tag, total int, err error) {
	m := dao.Tag.Ctx(ctx)

	if name != "" {
		m = m.WhereLike(dao.Tag.Columns().Name, "%"+name+"%")
	}
	if slug != "" {
		m = m.WhereLike(dao.Tag.Columns().Slug, "%"+slug+"%")
	}

	total, err = m.Count()
	if err != nil {
		return
	}

	err = m.OrderDesc(dao.Tag.Columns().Id).Page(page, size).Scan(&list)
	return
}

func (s *sTag) GetOne(ctx context.Context, id int) (tag *entity.Tag, err error) {
	err = dao.Tag.Ctx(ctx).Where(dao.Tag.Columns().Id, id).Scan(&tag)
	return
}

func (s *sTag) GetBySlug(ctx context.Context, slug string) (tag *entity.Tag, err error) {
	err = dao.Tag.Ctx(ctx).Where(dao.Tag.Columns().Slug, slug).Scan(&tag)
	return
}

func (s *sTag) Delete(ctx context.Context, id int) (err error) {
	_, err = dao.Tag.Ctx(ctx).Where(dao.Tag.Columns().Id, id).Delete()
	return
}

